### PR TITLE
Csv from file

### PR DESCRIPTION
**Pull request by toryjleo**

### CODE REVIEW

**Comment by yiskylee** [comment]
This looks good and replacing the delimiter with ' ' is very smart! Couple of thoughts though:
1. Could you use std::string instead of char for the delimiter? Or did you use char instead to avoid some problems caused by using string?
2. Could you in you tests also test csv files that you manually wrote, for example a.csv, and then pass in ',' instead of ' ' to see if it works? 
3. We should only allow user to pass in a delimiter that is one of ' ', '\t' and ','. Could you add a checker in FromFile that reports errors when users are trying to pass in other characters?
4. When the delimiter is ' ', it is not efficient to do the std::replace, could you add a condition that check that?
5. When a user passes in a csv file but uses delimiter ' ' or '\t' or vice versa, could you add some checker to the code so that it reports errors? 
6. Please also add Doxygen comments to FromFile()

**Comment by toryjleo** [comment]
Hey Xiangyu, it does have to be char; std::string didn't want to play nice. Also, does it matter if the delimiter is ' ' or ' \t'? It is my belief that they can both be read in the same way.

**Comment by toryjleo** [comment]
Hey, Xiangyu, I would appreciate it if you took at my code.

File: cpp/test/util_test/from_file_test.cc [hunk 1]
```diff
@@ -90,17 +99,8 @@ TYPED_TEST(FromFileTest, CSVTest) {
...
-        this->Filer(
-            "../test/data_for_test/matrix_not_exist.txt", 2, 2, ' ');
-      }
-      , "Cannot open file .*, exiting...");
-}
-
-TYPED_TEST(FromFileTest, IfFileExistsRowsAndColsMethod) {
-  this->Filer("../test/data_for_test/matrix_2_2.txt", 2, 2, ' ');
+TYPED_TEST(FromFileTest, CSVTest) {
+  this->Filer("../test/data_for_test/matrix_2_2_csv.txt", ",");
```
Comment (yiskylee) [inline cpp/test/util_test/from_file_test.cc]: could you rename all *_csv.txt files to *.csv files (e.g. matrix_2_2_csv.txt -> matrix_2_2.csv)? The extension should be .csv for csv files.

File: cpp/test/util_test/from_file_test.cc [hunk 2]
```diff
@@ -112,7 +112,7 @@ TYPED_TEST(FromFileTest, IfFileExistsRowsAndColsMethod) {
 }
 
 TYPED_TEST(FromFileTest, CSVRowsAndColsMethodTest) {
-  this->Filer("../test/data_for_test/matrix_2_2_csv.txt", 2, 2, ',');
+  this->Filer("../test/data_for_test/matrix_2_2_csv.txt", 2, 2, ",");
```
Comment (yiskylee) [inline cpp/test/util_test/from_file_test.cc]: Could you make a directory named FromFile under ../test/data_for_test/ and put the corresponding test data files in there?

**Comment by toryjleo** [comment]
I made those changes

File: cpp/include/util.h [hunk 3]
```diff
@@ -114,8 +127,21 @@ Matrix<T> FromFile(const std::string &input_file_path,
       if (line.find_first_not_of(' ') == std::string::npos) {
```
Comment (yiskylee) [inline cpp/include/util.h]: For one-liners in the if block, it is suggested to skip the curly braces

**Comment by yiskylee** [comment]
Looks good! Just couple of last changes:
1. I commented on one of if statement in the util.h, please don't use {} when there is only one line after if.
2. Please rename colsinrow to cols_in_row
3. line 149 while ( stream >> coef ), please delete the leading and trailing spaces inside ()
4. Please make a pull from upstream again. Previously we cannot see the doxy for FromFile because doxygen doesn't generate comments for namespaces. We have fixed that recently.

**Comment by toryjleo** [comment]
It is done.

**Comment by yiskylee** [comment]
This looks great! Thank you!

**Pull request merged**